modules/javafx.graphics/src/main/java/javafx/scene/image/PixelBuffer.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javafx.scene.image;
+
+import com.sun.javafx.geom.Rectangle;
+import com.sun.javafx.tk.Toolkit;
+import javafx.geometry.Rectangle2D;
+import javafx.util.Callback;
+
+import java.lang.ref.WeakReference;
+import java.nio.Buffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Objects;
+
+/**
+ * A class to provide a shared pixel data using {@link java.nio.Buffer} for {@code WritableImage}s.
+ * {@code PixelBuffer} supports only {@code PixelFormat.INT_ARGB_PRE}
+ * and {@code PixelFormat.BYTE_BGRA_PRE} pixel formats.
+ * A {@code PixelBuffer} can only be drawn using a {@code WritableImage}.
+ * Multiple {@code WritableImage}s and applications can share a {@code PixelBuffer}.
+ * <p>
+ * Pseudo code to create {@code PixelBuffer}:
+ * <pre>{@code
+ *
+ * // Creating a PixelBuffer using PixelFormat.BYTE_BGRA_PRE
+ * ByteBuffer byteBuffer = ByteBuffer.allocateDirect(width * height * 4);
+ * PixelFormat<ByteBuffer> pixelFormat = PixelFormat.getByteBgraPreInstance();
+ * PixelBuffer pixelBuffer = new PixelBuffer(width, height, byteBuffer, pixelFormat);
+ * Image img = new WritableImage(pixelBuffer);
+ *
+ * // Creating a PixelBuffer using PixelFormat.INT_ARGB_PRE
+ * IntBuffer intBuffer = IntBuffer.allocate(width * height);
+ * PixelFormat<IntBuffer> pixelFormat = PixelFormat.getIntArgbPreInstance();
+ * PixelBuffer pixelBuffer = new PixelBuffer(width, height, intBuffer, pixelFormat);
+ * Image img = new WritableImage(pixelBuffer);
+ * }</pre>
+ *
```

Comment:
Good catch. I noticed this a couple weeks ago while looking at the Draft CSR. I made a mental note to add a review comment, but forgot by the time I got back to it.

Reply:
Oh, and the missing `@param <T>` in `PixelFormat` should be fixed separately since that class is otherwise not modified by this enhancement -- perhaps it could be fixed as part of [JDK-8221377](https://bugs.openjdk.java.net/browse/JDK-8221377)?

Reply:
I can fix `PixelFormat` in JDK-8221377. By the way, there are 2 issues that require your attention there. Most of the other issues are already fixed locally.

Reply:
Added `@param <T>`  and  `PixelFormat`  class can be fixed in JDK-8221377, resolving this conversation